routing/payment_lifecycle.go

```diff
@@ -26,189 +28,524 @@ func (e errNoRoute) Error() string {
-		// switch.
-		errorDecryptor := &htlcswitch.SphinxErrorDecrypter{
-			OnionErrorDecrypter: sphinx.NewOnionErrorDecrypter(p.circuit),
+			return [32]byte{}, nil, fmt.Errorf("No successful " +
+				"attempts nor failure found!")
 		}
 
-		// Now ask the switch to return the result of the payment when
-		// available.
-		resultChan, err := p.router.cfg.Payer.GetPaymentResult(
-			p.attempt.AttemptID, p.payment.PaymentHash, errorDecryptor,
-		)
-		switch {
+		remValue := p.totalAmount - sentAmt
+
+	ShardLoop:
+		// While we still have amount to send, and we didn't aleady
+		// encounter a success or parmanent failure, send another
+		// shard.
+		for remValue > 0 && !(success || failure != nil) {
+			// Before we attempt this next payment, we'll check to see if either
+			// we've gone past the payment attempt timeout, or the router is
+			// exiting. In either case, we'll stop this payment attempt short. If a
+			// timeout is not applicable, timeoutChan will be nil.
+			select {
+			case <-p.timeoutChan:
+				errStr := fmt.Sprintf("payment attempt not completed " +
+					"before timeout")
+
+				terminalFailure = &paymentFailure{
+					failureCode: channeldb.FailureReasonTimeout,
+					err:         newErr(ErrPaymentAttemptTimeout, errStr),
+				}
+				break ShardLoop
 
-		// If this attempt ID is unknown to the Switch, it means it was
-		// never checkpointed and forwarded by the switch before a
-		// restart. In this case we can safely send a new payment
-		// attempt, and wait for its result to be available.
-		case err == htlcswitch.ErrPaymentIDNotFound:
-			log.Debugf("Payment ID %v for hash %x not found in "+
-				"the Switch, retrying.", p.attempt.AttemptID,
-				p.payment.PaymentHash)
+			// Fall through if we haven't hit our time limit.
+			default:
+			}
 
-			err = p.failAttempt(err)
+			// Create a new payment attempt from the given payment session.
+			rt, err := p.paySession.RequestRoute(
+				remValue, uint32(p.currentHeight),
+			)
 			if err != nil {
-				return [32]byte{}, nil, err
-			}
+				log.Warnf("Failed to find route for payment %x: %v",
+					p.paymentHash, err)
 
-			// Reset the attempt to indicate we want to make a new
-			// attempt.
-			p.attempt = nil
-			continue
+				// Convert error to payment-level failure.
+				routeFailure = &paymentFailure{
+					failureCode: errorToPaymentFailure(err),
+					err:         err,
+				}
 
-		// A critical, unexpected error was encountered.
-		case err != nil:
-			log.Errorf("Failed getting result for attemptID %d "+
-				"from switch: %v", p.attempt.AttemptID, err)
+				// If there was an error already recorded for this
+				// payment, we'll return that.
+				if lastError != nil {
+					routeFailure.err = errNoRoute{lastError: lastError}
+				}
 
-			return [32]byte{}, nil, err
+				break ShardLoop
+			}
+
+			// Launch a shard along the route, getting the outcome
+			// delivered on the channel.
+			shardHandler.launchAndCollect(
+				rt, shardOutcomes, criticalErrs,
+			)
```

Comment:
Would it be better to check shard outcomes after every pathfinding operation? Maybe the remaining amount is more up to date then.

Reply:
Good idea, done!